### DISCUSSION TITLE
Source controller continuously recreating helm chart packages

### DISCUSSION THREAD

**Discussion by matt-woodruff-f3** [fluxcd/flux2#2225] [Q&A]
Hello!

On my current project we were experiencing large helm chart reconciliation times and discovered the problem to be slow disk writes caused by other software we're running. 

After looking deeper into the problem we noticed when helm charts are packaged from a git repository they're always repackaged at the given interval even if the packages have not changed. In contrast to pulling charts from helm repositories, we can see this doesn't occur. We've also looked through the source code and believe this to be intentional.

Can I confirm this is desired behaviour and if there are other ways we can prevent the helm chart to be repackaged each time? The bottleneck is untar-ing the git repository containing all our charts, there are many charts (200) and this operation takes time. Our source controller also has a constant queue and can take 4-5 mins to reconcile something new.

Some options we've thought about
- Increasing source controller concurrency. This provides no benefit as the bottleneck is disk writes.
- Creating a gitrepository for each chart and ignore all the irrelevant charts to reduce file writes when untar-ing. We have around 200 helm releases in some clusters so we wanted to avoid this approach.
- Setting a larger interval value for the helm chart as they don't change often. This could eliminate our source controller queue but would we have to wait until the next interval to see chart changes? Or maybe a change in git repo would trigger a helm chart repackage? 

In the future we will look to use our own helm repository but for next month or so a git repository will be our chart source.

Thanks!

**Comment by kingdonb** [comment]
> when helm charts are packaged from a git repository they're always repackaged at the given interval even if the packages have not changed

Always repackaged at the interval, or always repackaged with each new commit? I would expect the latter but not the former.

Is this issue reflecting that the artifact is reconciled into a new HelmChart every time there is a new commit? Or every time the reconcile interval is reached?

(Yeah this is a problem, either way...)

The source controller serves up whole .tar.gz files of whatever was in the source. So if you have 200 charts served from the same gitrepo source, then each new commit has to be downloaded by each downstream individually (whole, with 199 neighbors intact) and extracted to figure out if there was any change.

Serving up helm charts from a gitrepository is something that has potential for performance problems for many reasons.

You may see better performance after upgrading to Flux 0.26, which includes some important Helm fixes for upstream issues related to memory leaks and has allowed a lot of folks to use Helm Controller under reasonable memory conditions when they were likely to experience OOM errors before.

But you have pegged it, the issue is disk writes; the charts need to be separated.

If the chart is really repackaged at every interval regardless of whether there has been any new commit or not (this is irrespective of whether or not the chart itself has changed; if there is a new commit, Flux has no way to compare it against the previous version and de-duplicate, but if there hasn't been a new commit SHA, I don't think it would update), if it is doing that, I think that is a bug.

Would you be willing to write up some concrete minimal reproduce instruction steps if you're still struggling with this particular issue in Flux 0.26+? File under Helm Controller I think (or Source Controller, we can move it if needed...) We're more and less likely to be able to do anything about parts of this issue; if you're getting a new chart artifact every interval with no changes, that's certainly a bug.
